<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork865-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo"  />
    </a>
</p>


# **SpaceX  Falcon 9 first stage Landing Prediction**


# Lab 1: Collecting the data


Estimated time needed: **45** minutes


In this capstone, we will predict if the Falcon 9 first stage will land successfully. SpaceX advertises Falcon 9 rocket launches on its website with a cost of 62 million dollars; other providers cost upward of 165 million dollars each, much of the savings is because SpaceX can reuse the first stage. Therefore if we can determine if the first stage will land, we can determine the cost of a launch. This information can be used if an alternate company wants to bid against SpaceX for a rocket launch. In this lab, you will collect and make sure the data is in the correct format from an API. The following is an example of a successful and launch.


![](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/lab_v2/images/landing_1.gif)


Several examples of an unsuccessful landing are shown here:


![](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/lab_v2/images/crash.gif)


Most unsuccessful landings are planned. Space X performs a controlled landing in the oceans. 


## Objectives


In this lab, you will make a get request to the SpaceX API. You will also do some basic data wrangling and formating. 

- Request to the SpaceX API
- Clean the requested data


----


## Import Libraries and Define Auxiliary Functions


We will import the following libraries into the lab


In [1]:
# Requests allows us to make HTTP requests which we will use to get data from an API
import requests
# Pandas is a software library written for the Python programming language for data manipulation and analysis.
import pandas as pd
# NumPy is a library for the Python programming language, adding support for large, multi-dimensional arrays and matrices, along with a large collection of high-level mathematical functions to operate on these arrays
import numpy as np
# Datetime is a library that allows us to represent dates
import datetime

# Setting this option will print all collumns of a dataframe
pd.set_option('display.max_columns', None)
# Setting this option will print all of the data in a feature
pd.set_option('display.max_colwidth', None)

Below we will define a series of helper functions that will help us use the API to extract information using identification numbers in the launch data.

From the <code>rocket</code> column we would like to learn the booster name.


In [2]:
# Takes the dataset and uses the rocket column to call the API and append the data to the list
def getBoosterVersion(data):
    for x in data['rocket']:
       if x:
        response = requests.get("https://api.spacexdata.com/v4/rockets/"+str(x)).json()
        BoosterVersion.append(response['name'])

From the <code>launchpad</code> we would like to know the name of the launch site being used, the logitude, and the latitude.


In [3]:
# Takes the dataset and uses the launchpad column to call the API and append the data to the list
def getLaunchSite(data):
    for x in data['launchpad']:
       if x:
         response = requests.get("https://api.spacexdata.com/v4/launchpads/"+str(x)).json()
         Longitude.append(response['longitude'])
         Latitude.append(response['latitude'])
         LaunchSite.append(response['name'])

From the <code>payload</code> we would like to learn the mass of the payload and the orbit that it is going to.


In [4]:
# Takes the dataset and uses the payloads column to call the API and append the data to the lists
def getPayloadData(data):
    for load in data['payloads']:
       if load:
        response = requests.get("https://api.spacexdata.com/v4/payloads/"+load).json()
        PayloadMass.append(response['mass_kg'])
        Orbit.append(response['orbit'])

From <code>cores</code> we would like to learn the outcome of the landing, the type of the landing, number of flights with that core, whether gridfins were used, wheter the core is reused, wheter legs were used, the landing pad used, the block of the core which is a number used to seperate version of cores, the number of times this specific core has been reused, and the serial of the core.


In [5]:
# Takes the dataset and uses the cores column to call the API and append the data to the lists
def getCoreData(data):
    for core in data['cores']:
            if core['core'] != None:
                response = requests.get("https://api.spacexdata.com/v4/cores/"+core['core']).json()
                Block.append(response['block'])
                ReusedCount.append(response['reuse_count'])
                Serial.append(response['serial'])
            else:
                Block.append(None)
                ReusedCount.append(None)
                Serial.append(None)
            Outcome.append(str(core['landing_success'])+' '+str(core['landing_type']))
            Flights.append(core['flight'])
            GridFins.append(core['gridfins'])
            Reused.append(core['reused'])
            Legs.append(core['legs'])
            LandingPad.append(core['landpad'])

Now let's start requesting rocket launch data from SpaceX API with the following URL:


In [6]:
spacex_url="https://api.spacexdata.com/v4/launches/past"

In [7]:
response = requests.get(spacex_url)

Check the content of the response


In [ ]:
print(response.content)

You should see the response contains massive information about SpaceX launches. Next, let's try to discover some more relevant information for this project.


### Task 1: Request and parse the SpaceX launch data using the GET request


To make the requested JSON results more consistent, we will use the following static response object for this project:


In [9]:
static_json_url='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/API_call_spacex_api.json'

We should see that the request was successfull with the 200 status response code


In [10]:
response.status_code

200

Now we decode the response content as a Json using <code>.json()</code> and turn it into a Pandas dataframe using <code>.json_normalize()</code>


In [11]:
# Use json_normalize meethod to convert the json result into a dataframe
jsonResponse = response.json()
data = pd.json_normalize(jsonResponse)

Using the dataframe <code>data</code> print the first 5 rows


In [12]:
# Get the head of the dataframe
data.head(1)

static_fire_date_utc  static_fire_date_unix    net  window  \
0  2006-03-17T00:00:00.000Z           1.142554e+09  False     0.0   

                     rocket success  \
0  5e9d0d95eda69955f709d1eb   False   

                                                              failures  \
0  [{'time': 33, 'altitude': None, 'reason': 'merlin engine failure'}]   

                                            details crew ships capsules  \
0  Engine failure at 33 seconds and loss of vehicle   []    []       []   

                     payloads                 launchpad  flight_number  \
0  [5eb0e4b5b6c3bb0006eeb1e1]  5e9e4502f5090995de566f86              1   

        name                  date_utc   date_unix                 date_local  \
0  FalconSat  2006-03-24T22:30:00.000Z  1143239400  2006-03-25T10:30:00+12:00   

  date_precision  upcoming  \
0           hour     False   

                                                                                                                                                                                              cores  \
0  [{'core': '5e9e289df35918033d3b2623', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}]   

   auto_update    tbd launch_library_id                        id  \
0         True  False              None  5eb87cd9ffd86e000604b32a   

  fairings.reused fairings.recovery_attempt fairings.recovered fairings.ships  \
0           False                     False              False             []   

                                 links.patch.small  \
0  https://images2.imgbox.com/94/f2/NN6Ph45r_o.png   

                                 links.patch.large links.reddit.campaign  \
0  https://images2.imgbox.com/5b/02/QcxHUb5V_o.png                  None   

  links.reddit.launch links.reddit.media links.reddit.recovery  \
0                None               None                  None   

  links.flickr.small links.flickr.original links.presskit  \
0                 []                    []           None   

                                 links.webcast links.youtube_id  \
0  https://www.youtube.com/watch?v=0a_00nJ_Y88      0a_00nJ_Y88   

                                                                  links.article  \
0  https://www.space.com/2196-spacex-inaugural-falcon-1-rocket-lost-launch.html   

                         links.wikipedia  fairings  
0  https://en.wikipedia.org/wiki/DemoSat       NaN

You will notice that a lot of the data are IDs. For example the rocket column has no information about the rocket just an identification number.

We will now use the API again to get information about the launches using the IDs given for each launch. Specifically we will be using columns <code>rocket</code>, <code>payloads</code>, <code>launchpad</code>, and <code>cores</code>.


In [13]:
# Lets take a subset of our dataframe keeping only the features we want and the flight number, and date_utc.
data = data[['rocket', 'payloads', 'launchpad', 'cores', 'flight_number', 'date_utc']]
data.head(1)

rocket                    payloads  \
0  5e9d0d95eda69955f709d1eb  [5eb0e4b5b6c3bb0006eeb1e1]   

                  launchpad  \
0  5e9e4502f5090995de566f86   

                                                                                                                                                                                              cores  \
0  [{'core': '5e9e289df35918033d3b2623', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}]   

   flight_number                  date_utc  
0              1  2006-03-24T22:30:00.000Z

In [14]:
# We will remove rows with multiple cores because those are falcon rockets with 2 extra rocket boosters and rows that have multiple payloads in a single rocket.
data = data[data['cores'].map(len)==1]
data = data[data['payloads'].map(len)==1]
data.head(1)

rocket                    payloads  \
0  5e9d0d95eda69955f709d1eb  [5eb0e4b5b6c3bb0006eeb1e1]   

                  launchpad  \
0  5e9e4502f5090995de566f86   

                                                                                                                                                                                              cores  \
0  [{'core': '5e9e289df35918033d3b2623', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}]   

   flight_number                  date_utc  
0              1  2006-03-24T22:30:00.000Z

In [15]:
# Since payloads and cores are lists of size 1 we will also extract the single value in the list and replace the feature.
data['cores'] = data['cores'].map(lambda x : x[0])
data['payloads'] = data['payloads'].map(lambda x : x[0])
data.head(1)

rocket                  payloads  \
0  5e9d0d95eda69955f709d1eb  5eb0e4b5b6c3bb0006eeb1e1   

                  launchpad  \
0  5e9e4502f5090995de566f86   

                                                                                                                                                                                            cores  \
0  {'core': '5e9e289df35918033d3b2623', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}   

   flight_number                  date_utc  
0              1  2006-03-24T22:30:00.000Z

In [16]:
# We also want to convert the date_utc to a datetime datatype and then extracting the date leaving the time
data['date'] = pd.to_datetime(data['date_utc']).dt.date
data.head(1)

rocket                  payloads  \
0  5e9d0d95eda69955f709d1eb  5eb0e4b5b6c3bb0006eeb1e1   

                  launchpad  \
0  5e9e4502f5090995de566f86   

                                                                                                                                                                                            cores  \
0  {'core': '5e9e289df35918033d3b2623', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}   

   flight_number                  date_utc        date  
0              1  2006-03-24T22:30:00.000Z  2006-03-24

In [17]:
# Using the date we will restrict the dates of the launches
data = data[data['date'] <= datetime.date(2020, 11, 13)]
data.head(1)

rocket                  payloads  \
0  5e9d0d95eda69955f709d1eb  5eb0e4b5b6c3bb0006eeb1e1   

                  launchpad  \
0  5e9e4502f5090995de566f86   

                                                                                                                                                                                            cores  \
0  {'core': '5e9e289df35918033d3b2623', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}   

   flight_number                  date_utc        date  
0              1  2006-03-24T22:30:00.000Z  2006-03-24

* From the <code>rocket</code> we would like to learn the booster name

* From the <code>payload</code> we would like to learn the mass of the payload and the orbit that it is going to

* From the <code>launchpad</code> we would like to know the name of the launch site being used, the longitude, and the latitude.

* From <code>cores</code> we would like to learn the outcome of the landing, the type of the landing, number of flights with that core, whether gridfins were used, whether the core is reused, whether legs were used, the landing pad used, the block of the core which is a number used to seperate version of cores, the number of times this specific core has been reused, and the serial of the core.

The data from these requests will be stored in lists and will be used to create a new dataframe.


In [18]:
#Global variables 
BoosterVersion = []
PayloadMass = []
Orbit = []
LaunchSite = []
Outcome = []
Flights = []
GridFins = []
Reused = []
Legs = []
LandingPad = []
Block = []
ReusedCount = []
Serial = []
Longitude = []
Latitude = []

These functions will apply the outputs globally to the above variables. Let's take a looks at <code>BoosterVersion</code> variable. Before we apply  <code>getBoosterVersion</code> the list is empty:


In [19]:
BoosterVersion

[]

Now, let's apply <code> getBoosterVersion</code> function method to get the booster version


In [20]:
# Call getBoosterVersion
getBoosterVersion(data)
data.head(1)

rocket                  payloads  \
0  5e9d0d95eda69955f709d1eb  5eb0e4b5b6c3bb0006eeb1e1   

                  launchpad  \
0  5e9e4502f5090995de566f86   

                                                                                                                                                                                            cores  \
0  {'core': '5e9e289df35918033d3b2623', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}   

   flight_number                  date_utc        date  
0              1  2006-03-24T22:30:00.000Z  2006-03-24

the list has now been update 


In [21]:
BoosterVersion[0:5]

['Falcon 1', 'Falcon 1', 'Falcon 1', 'Falcon 1', 'Falcon 9']

we can apply the rest of the  functions here:


In [22]:
# Call getLaunchSite
getLaunchSite(data)
LaunchSite[0:5]

['Kwajalein Atoll',
 'Kwajalein Atoll',
 'Kwajalein Atoll',
 'Kwajalein Atoll',
 'CCSFS SLC 40']

In [23]:
def getPayloadData(data):
    for load in data['payloads']:
       if load:
        response = requests.get("https://api.spacexdata.com/v4/payloads/"+load).json()
        PayloadMass.append(response['mass_kg'])
        Orbit.append(response['orbit'])

In [24]:
# Call getPayloadData
getPayloadData(data)
data.head(1)

rocket                  payloads  \
0  5e9d0d95eda69955f709d1eb  5eb0e4b5b6c3bb0006eeb1e1   

                  launchpad  \
0  5e9e4502f5090995de566f86   

                                                                                                                                                                                            cores  \
0  {'core': '5e9e289df35918033d3b2623', 'flight': 1, 'gridfins': False, 'legs': False, 'reused': False, 'landing_attempt': False, 'landing_success': None, 'landing_type': None, 'landpad': None}   

   flight_number                  date_utc        date  
0              1  2006-03-24T22:30:00.000Z  2006-03-24

In [25]:
# Call getCoreData
getCoreData(data)

Finally lets construct our dataset using the data we have obtained. We we combine the columns into a dictionary.


In [26]:
launch_dict = {'FlightNumber': list(data['flight_number']),
'Date': list(data['date']),
'BoosterVersion':BoosterVersion,
'PayloadMass':PayloadMass,
'Orbit':Orbit,
'LaunchSite':LaunchSite,
'Outcome':Outcome,
'Flights':Flights,
'GridFins':GridFins,
'Reused':Reused,
'Legs':Legs,
'LandingPad':LandingPad,
'Block':Block,
'ReusedCount':ReusedCount,
'Serial':Serial,
'Longitude': Longitude,
'Latitude': Latitude}


Then, we need to create a Pandas data frame from the dictionary launch_dict.


In [27]:
# Create a data from launch_dict
df2= pd.DataFrame.from_dict(launch_dict, orient='index')
dfT = df2.T
dfT

FlightNumber        Date BoosterVersion PayloadMass Orbit       LaunchSite  \
0             1  2006-03-24       Falcon 1          20   LEO  Kwajalein Atoll   
1             2  2007-03-21       Falcon 1        None   LEO  Kwajalein Atoll   
2             4  2008-09-28       Falcon 1         165   LEO  Kwajalein Atoll   
3             5  2009-07-13       Falcon 1         200   LEO  Kwajalein Atoll   
4             6  2010-06-04       Falcon 9        None   LEO     CCSFS SLC 40   
..          ...         ...            ...         ...   ...              ...   
89          102  2020-09-03       Falcon 9       15600  VLEO       KSC LC 39A   
90          103  2020-10-06       Falcon 9       15600  VLEO       KSC LC 39A   
91          104  2020-10-18       Falcon 9       15600  VLEO       KSC LC 39A   
92          105  2020-10-24       Falcon 9       15600  VLEO     CCSFS SLC 40   
93          106  2020-11-05       Falcon 9        3681   MEO     CCSFS SLC 40   

      Outcome Flights GridFins Reused   Legs                LandingPad Block  \
0   None None       1    False  False  False                      None  None   
1   None None       1    False  False  False                      None  None   
2   None None       1    False  False  False                      None  None   
3   None None       1    False  False  False                      None  None   
4   None None       1    False  False  False                      None     1   
..        ...     ...      ...    ...    ...                       ...   ...   
89  True ASDS       2     True   True   True  5e9e3032383ecb6bb234e7ca     5   
90  True ASDS       3     True   True   True  5e9e3032383ecb6bb234e7ca     5   
91  True ASDS       6     True   True   True  5e9e3032383ecb6bb234e7ca     5   
92  True ASDS       3     True   True   True  5e9e3033383ecbb9e534e7cc     5   
93  True ASDS       1     True  False   True  5e9e3032383ecb6bb234e7ca     5   

   ReusedCount    Serial   Longitude   Latitude  
0            0  Merlin1A  167.743129   9.047721  
1            0  Merlin2A  167.743129   9.047721  
2            0  Merlin2C  167.743129   9.047721  
3            0  Merlin3C  167.743129   9.047721  
4            0     B0003  -80.577366  28.561857  
..         ...       ...         ...        ...  
89          12     B1060  -80.603956  28.608058  
90          13     B1058  -80.603956  28.608058  
91          12     B1051  -80.603956  28.608058  
92          12     B1060  -80.577366  28.561857  
93           8     B1062  -80.577366  28.561857  

[94 rows x 17 columns]

Show the summary of the dataframe


In [28]:
# Show the head of the dataframe
dfT.head(3)

FlightNumber        Date BoosterVersion PayloadMass Orbit       LaunchSite  \
0            1  2006-03-24       Falcon 1          20   LEO  Kwajalein Atoll   
1            2  2007-03-21       Falcon 1        None   LEO  Kwajalein Atoll   
2            4  2008-09-28       Falcon 1         165   LEO  Kwajalein Atoll   

     Outcome Flights GridFins Reused   Legs LandingPad Block ReusedCount  \
0  None None       1    False  False  False       None  None           0   
1  None None       1    False  False  False       None  None           0   
2  None None       1    False  False  False       None  None           0   

     Serial   Longitude  Latitude  
0  Merlin1A  167.743129  9.047721  
1  Merlin2A  167.743129  9.047721  
2  Merlin2C  167.743129  9.047721

In [29]:
missing_data = dfT.isnull()
missing_data.head(5)

FlightNumber   Date  BoosterVersion  PayloadMass  Orbit  LaunchSite  \
0         False  False           False        False  False       False   
1         False  False           False         True  False       False   
2         False  False           False        False  False       False   
3         False  False           False        False  False       False   
4         False  False           False         True  False       False   

   Outcome  Flights  GridFins  Reused   Legs  LandingPad  Block  ReusedCount  \
0    False    False     False   False  False        True   True        False   
1    False    False     False   False  False        True   True        False   
2    False    False     False   False  False        True   True        False   
3    False    False     False   False  False        True   True        False   
4    False    False     False   False  False        True  False        False   

   Serial  Longitude  Latitude  
0   False      False     False  
1   False      False     False  
2   False      False     False  
3   False      False     False  
4   False      False     False

In [30]:
for column in missing_data.columns.values.tolist():
    print(column)
    print (missing_data[column].value_counts())
    print("")

FlightNumber
False    94
Name: FlightNumber, dtype: int64

Date
False    94
Name: Date, dtype: int64

BoosterVersion
False    94
Name: BoosterVersion, dtype: int64

PayloadMass
False    88
True      6
Name: PayloadMass, dtype: int64

Orbit
False    94
Name: Orbit, dtype: int64

LaunchSite
False    94
Name: LaunchSite, dtype: int64

Outcome
False    94
Name: Outcome, dtype: int64

Flights
False    94
Name: Flights, dtype: int64

GridFins
False    94
Name: GridFins, dtype: int64

Reused
False    94
Name: Reused, dtype: int64

Legs
False    94
Name: Legs, dtype: int64

LandingPad
False    64
True     30
Name: LandingPad, dtype: int64

Block
False    90
True      4
Name: Block, dtype: int64

ReusedCount
False    94
Name: ReusedCount, dtype: int64

Serial
False    94
Name: Serial, dtype: int64

Longitude
False    94
Name: Longitude, dtype: int64

Latitude
False    94
Name: Latitude, dtype: int64



### Task 2: Filter the dataframe to only include `Falcon 9` launches


Finally we will remove the Falcon 1 launches keeping only the Falcon 9 launches. Filter the data dataframe using the <code>BoosterVersion</code> column to only keep the Falcon 9 launches. Save the filtered data to a new dataframe called <code>data_falcon9</code>.


In [34]:
# Hint data['BoosterVersion']!='Falcon 1'
#data_falcon9= data['BoosterVersion']!='Falcon 1'
data_falcon9 = dfT[dfT.BoosterVersion != "Falcon 1"]
data_falcon9.head(1)

FlightNumber        Date BoosterVersion PayloadMass Orbit    LaunchSite  \
4            6  2010-06-04       Falcon 9        None   LEO  CCSFS SLC 40   

     Outcome Flights GridFins Reused   Legs LandingPad Block ReusedCount  \
4  None None       1    False  False  False       None     1           0   

  Serial  Longitude   Latitude  
4  B0003 -80.577366  28.561857

Now that we have removed some values we should reset the FlgihtNumber column


In [50]:
data_falcon9.loc[:,'FlightNumber'] = list(range(1, data_falcon9.shape[0]+1))
data_falcon9.tail(10)

/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/pandas/core/indexing.py:1773: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(ilocs[0], value, pi)


FlightNumber        Date BoosterVersion  PayloadMass Orbit    LaunchSite  \
84            81  2020-06-13       Falcon 9      15600.0  VLEO  CCSFS SLC 40   
85            82  2020-06-30       Falcon 9       3880.0   MEO  CCSFS SLC 40   
86            83  2020-07-20       Falcon 9       6124.0   GEO  CCSFS SLC 40   
87            84  2020-08-18       Falcon 9      15600.0  VLEO  CCSFS SLC 40   
88            85  2020-08-30       Falcon 9       1600.0   SSO  CCSFS SLC 40   
89            86  2020-09-03       Falcon 9      15600.0  VLEO    KSC LC 39A   
90            87  2020-10-06       Falcon 9      15600.0  VLEO    KSC LC 39A   
91            88  2020-10-18       Falcon 9      15600.0  VLEO    KSC LC 39A   
92            89  2020-10-24       Falcon 9      15600.0  VLEO  CCSFS SLC 40   
93            90  2020-11-05       Falcon 9       3681.0   MEO  CCSFS SLC 40   

      Outcome Flights GridFins Reused  Legs                LandingPad Block  \
84  True ASDS       3     True   True  True  5e9e3032383ecb6bb234e7ca     5   
85  True ASDS       1     True  False  True  5e9e3033383ecbb9e534e7cc     5   
86  True ASDS       2     True   True  True  5e9e3033383ecbb9e534e7cc     5   
87  True ASDS       6     True   True  True  5e9e3032383ecb6bb234e7ca     5   
88  True RTLS       4     True   True  True  5e9e3032383ecb267a34e7c7     5   
89  True ASDS       2     True   True  True  5e9e3032383ecb6bb234e7ca     5   
90  True ASDS       3     True   True  True  5e9e3032383ecb6bb234e7ca     5   
91  True ASDS       6     True   True  True  5e9e3032383ecb6bb234e7ca     5   
92  True ASDS       3     True   True  True  5e9e3033383ecbb9e534e7cc     5   
93  True ASDS       1     True  False  True  5e9e3032383ecb6bb234e7ca     5   

   ReusedCount Serial  Longitude   Latitude  
84           5  B1059 -80.577366  28.561857  
85          12  B1060 -80.577366  28.561857  
86          13  B1058 -80.577366  28.561857  
87           9  B1049 -80.577366  28.561857  
88           5  B1059 -80.577366  28.561857  
89          12  B1060 -80.603956  28.608058  
90          13  B1058 -80.603956  28.608058  
91          12  B1051 -80.603956  28.608058  
92          12  B1060 -80.577366  28.561857  
93           8  B1062 -80.577366  28.561857

In [41]:
avg_PayloadMass = data_falcon9["PayloadMass"].astype("float").mean(axis=0)
ravg_PayloadMass = round(avg_PayloadMass,)

In [42]:
data_falcon9["PayloadMass"].replace(np.nan, ravg_PayloadMass, inplace=True)
data_falcon9.head(1)

/home/jupyterlab/conda/envs/python/lib/python3.7/site-packages/pandas/core/generic.py:6619: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return self._update_inplace(result)


FlightNumber        Date BoosterVersion  PayloadMass Orbit    LaunchSite  \
4             1  2010-06-04       Falcon 9       6124.0   LEO  CCSFS SLC 40   

     Outcome Flights GridFins Reused   Legs LandingPad Block ReusedCount  \
4  None None       1    False  False  False       None     1           0   

  Serial  Longitude   Latitude  
4  B0003 -80.577366  28.561857

In [43]:
missing_data2 = data_falcon9.isnull()
for column in missing_data2.columns.values.tolist():
    print(column)
    print (missing_data2[column].value_counts())
    print("")

FlightNumber
False    90
Name: FlightNumber, dtype: int64

Date
False    90
Name: Date, dtype: int64

BoosterVersion
False    90
Name: BoosterVersion, dtype: int64

PayloadMass
False    90
Name: PayloadMass, dtype: int64

Orbit
False    90
Name: Orbit, dtype: int64

LaunchSite
False    90
Name: LaunchSite, dtype: int64

Outcome
False    90
Name: Outcome, dtype: int64

Flights
False    90
Name: Flights, dtype: int64

GridFins
False    90
Name: GridFins, dtype: int64

Reused
False    90
Name: Reused, dtype: int64

Legs
False    90
Name: Legs, dtype: int64

LandingPad
False    64
True     26
Name: LandingPad, dtype: int64

Block
False    90
Name: Block, dtype: int64

ReusedCount
False    90
Name: ReusedCount, dtype: int64

Serial
False    90
Name: Serial, dtype: int64

Longitude
False    90
Name: Longitude, dtype: int64

Latitude
False    90
Name: Latitude, dtype: int64



## Data Wrangling


We can see below that some of the rows are missing values in our dataset.


In [44]:
data_falcon9.isnull().sum()

FlightNumber       0
Date               0
BoosterVersion     0
PayloadMass        0
Orbit              0
LaunchSite         0
Outcome            0
Flights            0
GridFins           0
Reused             0
Legs               0
LandingPad        26
Block              0
ReusedCount        0
Serial             0
Longitude          0
Latitude           0
dtype: int64

Before we can continue we must deal with these missing values. The <code>LandingPad</code> column will retain None values to represent when landing pads were not used.


### Task 3: Dealing with Missing Values


Calculate below the mean for the <code>PayloadMass</code> using the <code>.mean()</code>. Then use the mean and the <code>.replace()</code> function to replace `np.nan` values in the data with the mean you calculated.


You should see the number of missing values of the <code>PayLoadMass</code> change to zero.


In [51]:
data_falcon9.to_csv('dataset_part_1.csv', index=False)

Now we should have no missing values in our dataset except for in <code>LandingPad</code>.


We can now export it to a <b>CSV</b> for the next section,but to make the answers consistent, in the next lab we will provide data in a pre-selected date range. 


<code>data_falcon9.to_csv('dataset_part_1.csv', index=False)</code>


## Authors


<a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDS0321ENSkillsNetwork865-2022-01-01">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD. 


## Change Log


|Date (YYYY-MM-DD)|Version|Changed By|Change Description|
|-|-|-|-|
|2020-09-20|1.1|Joseph|get result each time you run|
|2020-09-20|1.1|Azim |Created Part 1 Lab using SpaceX API|
|2020-09-20|1.0|Joseph |Modified Multiple Areas|


Copyright © 2021 IBM Corporation. All rights reserved.
